### PR TITLE
Cal/sparse set

### PR DESCRIPTION
**Pull request by McCallisterRomer**
-Removed ComponentSystem
-New type Registry owns storage for each component type + the EntitySystem
-Changed EntityHandle to uint64_t made from concatenating multiple smaller integers
-All places where component pointers were caches were changed to lookups using EntityHandle
-Added a to add staging area for EntitySystem
-HandleManager can recycle old handles
-Fixed bug in CollisionBenchmark where adding/removing in the middle of rendering could crash 
-Transform unit tests are intentionally commented out - they are broken with new changes, and Transform is likely changing soon

### CODE REVIEW

File: nc/source/ecs/ColliderTree.h [hunk 1]
```diff
@@ -15,6 +15,7 @@ namespace nc
     {
         using allow_trivial_destruction = std::true_type;
         using sort_dense_storage_by_address = std::true_type;
+        // ignore callbacks
```
Comment (jaremieromer) [inline nc/source/ecs/ColliderTree.h:18]: Is this a todo?

File: nc/source/ecs/Registry.h [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        auto requiredSize = storage.componentPool.size() + additionalRequiredCount;
+        storage.componentPool.reserve(requiredSize);
+        storage.entityPool.reserve(requiredSize);
+    }
+
+    template<class... Ts>
+    template<class T>
+    void Registry<Ts...>::RegisterOnAddCallback(typename SystemCallbacks<T>::on_add_type func)
+    {
+        static_assert(StoragePolicy<T>::requires_on_add_callback::value, "Cannot register an OnAdd callback unless specified in the StoragePolicy");
```
Comment (jaremieromer) [inline nc/source/ecs/Registry.h:299]: It's neat that these are compile-time validated.

File: project/source/spawn_test/SpawnTest.cpp [hunk 3]
```diff
@@ -68,7 +68,7 @@ namespace nc::sample
         staticCubeSpawner->SetPrefab(prefab::Resource::CubeRed);
         staticCubeSpawner->Spawn(5);
 
-        // Fixed interval spawner for moving cubes
+        //Fixed interval spawner for moving cubes
```
Comment (jaremieromer) [inline project/source/spawn_test/SpawnTest.cpp:71]: Why is this changed? I prefer one space, and the others in this file are one space.

**Review by jaremieromer — APPROVED** [review]

File: project/source/spawn_test/SpawnTest.cpp (hunk 3 above)
Comment (McCallisterRomer) [reply project/source/spawn_test/SpawnTest.cpp:71]: no idea

File: nc/source/ecs/ColliderTree.h (hunk 1 above)
Comment (McCallisterRomer) [reply nc/source/ecs/ColliderTree.h:18]: No, just commenting that I've intentionally left out aliases that won't be used.
